elements/pfe-clipboard/src/pfe-clipboard.js

```diff
@@ -0,0 +1,173 @@
+        slotClass: "pfe-clipboard__text--success",
+        slotId: "text--success"
+      }
+    };
+  }
+
+  constructor() {
+    super(PfeClipboard, { type: PfeClipboard.PfeType });
+
+    this._icon = this.shadowRoot.querySelector(`#icon`);
+    this._text = this.shadowRoot.querySelector(`#text`);
+  }
+
+  connectedCallback() {
+    super.connectedCallback();
+    // If you need to initialize any attributes, do that here
+
+    this.icon = this.querySelector(`[slot="${this.tag}--icon"]`);
+    this.text = this.querySelector(`[slot="${this.tag}--text"]`);
+    this.textSuccess = this.querySelector(`[slot="${this.tag}--text--success"]`);
+
+    // Add a slotchange listener to the lightDOM trigger
+    // this.icon.addEventListener("slotchange", this._init);
+
+    // Add accessibility attributes to treat this element as a button
+    this._containerElement = this.shadowRoot.querySelector(".pfe-clipboard__container");
+    this._containerElement.setAttribute("role", "button");
+    this._containerElement.setAttribute("tabindex", "0");
+
+    // Since this element as the role of button we are going to listen
+    // for click and as well as 'enter' and 'space' commands to trigger
+    // the copy functionality
+    this._containerElement.addEventListener("click", this._clickHandler.bind(this));
+    this._containerElement.addEventListener("keydown", this._keydownHandler.bind(this));
+  }
+
+  disconnectedCallback() {
+    this._containerElement.removeEventListener("click", this._clickHandler.bind(this));
+    this._containerElement.removeEventListener("keydown", this._keydownHandler.bind(this));
+  }
+
+  _clickHandler(event) {
+    // Execute the copy to clipboard functionality
+    this.copyURLToClipboard()
+      .then(url => {
+        // Emit event that lets others know the user has "copied"
+        // the button. We are also going to include the url that was
+        // copied.
+        this.emitEvent(PfeClipboard.events.copied, {
+          detail: {
+            url
+          }
+        });
+        // Toggle the copied state for 3 seconds
+        this.setAttribute("copied", "");
+        setTimeout(() => {
+          this.removeAttribute("copied");
+        }, 3000);
+      })
+      .catch(error => {
+        this.warn(error);
+      });
+  }
+
+  // Listen for keyboard events and map them to their
+  // corresponding mouse events.
+  _keydownHandler(event) {
+    let key = event.key || event.keyCode;
+    switch (key) {
+      case "Enter" || 13:
+        this._clickHandler(event);
+        break;
+      case " " || 32:
+        // Prevent the browser from scolling when the user hits the space key
+        event.stopPropagation();
+        event.preventDefault();
+        this._clickHandler(event);
+        break;
+    }
+  }
+
+  /**
+   * Copy url to the user's system clipboard clipboard
+   * https://caniuse.com/mdn-api_navigator_clipboard
+   * @async
+   * @return {Promise<string>} url
+   */
+  copyURLToClipboard() {
```

Comment:
Awesome approach; love the use of the promise here

Reply:
I really like it too!  Promises for the win. 👍